test/jdk/java/util/Collections/DelegatingIteratorForEachRemaining.java

```diff
@@ -195,4 +196,41 @@ private static Map<String, Object> map() {
         Class clazz = entrySet.iterator().next().getClass();
         assertThrowingIterator(Collections.checkedSet(entrySet, clazz).iterator());
     }
+
+    /**
+     * Calls Collections.unmodifiableMap().entrySet().iterator().forEachRemaining() by passing
+     * that method a {@code null} action and expects that call to fail with a
+     * {@code NullPointerException}. Additionally, the test verifies that the such a (failed) call
+     * doesn't advance the iterator to the next entry.
+     */
+    @Test
+    public void testUnmodifiableForEachRemainingNPE() {
+        final Map<String, String> singleEntryMap = Collections.singletonMap("foo", "bar");
+        final Iterator<Map.Entry<String, String>> it = Collections.unmodifiableMap(singleEntryMap)
+                .entrySet().iterator();
+        Assert.assertTrue("Iterator unexpectedly doesn't have any entry", it.hasNext());
+        // pass null "action" and expect it to fail with NPE
+        Assert.assertThrows(NullPointerException.class, () -> it.forEachRemaining(null));
+        // verify the iterator didn't advance
+        Assert.assertTrue("Iterator unexpectedly doesn't have any entry", it.hasNext());
+    }
+
+    /**
+     * Calls Collections.checkedMap().entrySet().iterator().forEachRemaining() by passing
+     * that method a {@code null} action and expects that call to fail with a
+     * {@code NullPointerException}. Additionally, the test verifies that the such a (failed) call
+     * doesn't advance the iterator to the next entry.
+     */
+    @Test
+    public void testCheckedMapForEachRemainingNPE() {
+        final Map<String, String> checkedMap = Collections.checkedMap(new HashMap<>(), String.class,
+                String.class);
+        checkedMap.put("hello", "world");
+        final Iterator<Map.Entry<String, String>> it = checkedMap.entrySet().iterator();
+        Assert.assertTrue("Iterator unexpectedly doesn't have any entry", it.hasNext());
+        // pass null "action" and expect it to fail with NPE
+        Assert.assertThrows(NullPointerException.class, () -> it.forEachRemaining(null));
+        // verify the iterator didn't advance
+        Assert.assertTrue("Iterator unexpectedly doesn't have any entry", it.hasNext());
+    }
```

Comment:
The checks in Collections.java look good.

The tests can be simplified, I think. The reproducers in the original bug report wrap an empty map in either an unmodifiable map or a checked map, so you could just test them with entrySet().iterator().forEachRemaining(null). Those cases do nothing in the current JDK when they indeed should throw NPE.

Reply:
Hello Stuart, I have now updated the PR to simplify the test as suggested.